entity-service/src/main/kotlin/com/egm/stellio/entity/repository/QueryUtils.kt

```diff
@@ -191,4 +186,79 @@ object QueryUtils {
                }
             """.trimIndent()
         }
+
+    fun prepareQueryForAuthorizedEntitiesWithAuthentication(
+        queryParams: QueryParams,
+        offset: Int,
+        limit: Int
+    ): String {
+        val (_, expandedType, _, q, _) = queryParams
+        val matchEntityClause = buildMatchEntityClause(expandedType, prefix = "")
+        val authTerm = buildAuthTerm(q)
+        val matchAuthorizedEntitiesClause =
+            """
+            CALL {
+                MATCH (user)
+                WHERE user.id IN ${'$'}userAndGroupIds
+                WITH user
+                MATCH (user)-[]->()-[right:$authTerm]->$matchEntityClause
+                OPTIONAL MATCH $matchEntityClause-[:HAS_VALUE]->(prop:Property { name: '$AUTH_PROP_SAP' })
+                WHERE prop.value IN [
+                    '${AuthContextModel.SpecificAccessPolicy.AUTH_READ.name}'
+                ]
+                RETURN {entity: entity, right: right, specificAccessPolicy: prop.value} as entities 
+            }
+            WITH collect(distinct(entities)) as entities, count(distinct(entities)) as count 
+            """.trimIndent()
+
+        val pagingClause = if (limit == 0)
+            """
+            RETURN count
+            """.trimIndent()
+        else
+            """
+            RETURN entities, count
+            ORDER BY entities
+            SKIP $offset LIMIT $limit
+            """.trimIndent()
+
+        return """
+            $matchAuthorizedEntitiesClause
+            $pagingClause
+            """
+    }
+
+    fun prepareQueryForAuthorizedEntitiesWithoutAuthentication(
+        queryParams: QueryParams,
+        offset: Int,
+        limit: Int
+    ): String {
+        val (_, expandedType, _, _, _) = queryParams
+        val matchEntityClause = buildMatchEntityClause(expandedType)
+        val pagingClause = if (limit == 0)
+            """
+            RETURN count(entity) as count
+            """.trimIndent()
+        else
+            """
+            WITH collect(entity) as entities, count(entity) as count
+            RETURN {entity: entities, right: "rCanWrite"} as entities, count
```

Comment:
why `rCanWrite`? nothing about `specificAccessPolicy`?

Reply:
It's not `rCanWrite` but `rCanAdmin`. When we are admin it's necessary to display `specificAccessPolicy` ?